File: algo.py
def fcfs(requests, initial_head):
    sequence = [initial_head] + requests
    distance = 0
    for i in range(len(sequence) - 1):
        distance += abs(sequence[i+1] - sequence[i])
    return sequence, distance
def sstf(requests, initial_head):
    sequence = [initial_head]
    distance = 0
    pending = requests.copy()
    current_head = initial_head
    
    while pending:
        next_request = min(pending, key=lambda r: abs(r - current_head))
        distance += abs(next_request - current_head)
        sequence.append(next_request)
        current_head = next_request
        pending.remove(next_request)
    
    return sequence, distance

def scan(requests, initial_head,max_cylinder,direction="Up"):
    sequence = [initial_head]
    distance = 0
    current_head = initial_head
    sorted_requests = sorted(requests)

    left = [r for r in sorted_requests if r<initial_head]
    right = [r for r in sorted_requests if r>=initial_head]
    if direction =="Up":
        path = right + [max_cylinder-1] + left[::-1]
    else:
        path = left[::-1] + [0] + right
    for req in path:
        if req != current_head:
            sequence.append(req)
            distance += abs(req - current_head)
            current_head = req
    clean_seq = [sequence[0]]
    for i in sequence[1:]:
        if i != clean_seq[-1]:
            clean_seq.append(i)
    return clean_seq, distance

def cscan(requests, initial_head,max_cylinder,direction="Up"):
    sequence = [initial_head]
    distance = 0
    current_head = initial_head
    sorted_requests = sorted(requests)
    left = [r for r in sorted_requests if r<initial_head]
    right = [r for r in sorted_requests if r>=initial_head]
    if direction == "Up":
        path = right + [max_cylinder-1,0]+ left
    else:
        path = left[::-1] + [0,max_cylinder-1] + right[::-1]
    for req in path:
        if req != current_head:
            sequence.append(req)
            distance += abs(req - current_head)
            current_head = req
    clean_seq = [sequence[0]]
    for i in sequence[1:]:
        if i != clean_seq[-1]:
            clean_seq.append(i) 
    return clean_seq, distance

File: frontend/src/algo.js
/**
 * Disk Scheduling Algorithms — pure JavaScript implementation
 * (mirrors the C implementation in c_algo/disk_scheduling.c)
 *
 * Each function returns { sequence: number[], totalDistance: number }
 */

export function fcfs(requests, initialHead) {
  const sequence = [initialHead];
  let distance = 0;
  let current = initialHead;

  for (const req of requests) {
    distance += Math.abs(req - current);
    current = req;
    sequence.push(current);
  }

  return { sequence, totalDistance: distance };
}

export function sstf(requests, initialHead) {
  const sequence = [initialHead];
  let distance = 0;
  let current = initialHead;
  const pending = [...requests];
  const served = new Array(pending.length).fill(false);

  for (let done = 0; done < pending.length; done++) {
    let bestIdx = -1;
    let bestDist = Infinity;

    for (let i = 0; i < pending.length; i++) {
      if (!served[i]) {
        const d = Math.abs(pending[i] - current);
        if (d < bestDist) {
          bestDist = d;
          bestIdx = i;
        }
      }
    }

    served[bestIdx] = true;
    distance += bestDist;
    current = pending[bestIdx];
    sequence.push(current);
  }

  return { sequence, totalDistance: distance };
}

export function scan(requests, initialHead, maxCylinder, direction = 'Up') {
  const sorted = [...requests].sort((a, b) => a - b);
  const left = sorted.filter((r) => r < initialHead);
  const right = sorted.filter((r) => r >= initialHead);

  let path;
  if (direction === 'Up') {
    path = [...right, maxCylinder - 1, ...left.reverse()];
  } else {
    path = [...[...left].reverse(), 0, ...right];
  }

  const sequence = [initialHead];
  let distance = 0;
  let current = initialHead;

  for (const p of path) {
    if (p !== current) {
      distance += Math.abs(p - current);
      current = p;
      sequence.push(current);
    }
  }

  // deduplicate consecutive
  const clean = [sequence[0]];
  for (let i = 1; i < sequence.length; i++) {
    if (sequence[i] !== clean[clean.length - 1]) clean.push(sequence[i]);
  }

  return { sequence: clean, totalDistance: distance };
}

export function cscan(requests, initialHead, maxCylinder, direction = 'Up') {
  const sorted = [...requests].sort((a, b) => a - b);
  const left = sorted.filter((r) => r < initialHead);
  const right = sorted.filter((r) => r >= initialHead);

  let path;
  if (direction === 'Up') {
    path = [...right, maxCylinder - 1, 0, ...left];
  } else {
    path = [...[...left].reverse(), 0, maxCylinder - 1, ...[...right].reverse()];
  }

  const sequence = [initialHead];
  let distance = 0;
  let current = initialHead;

  for (const p of path) {
    if (p !== current) {
      distance += Math.abs(p - current);
      current = p;
      sequence.push(current);
    }
  }

  const clean = [sequence[0]];
  for (let i = 1; i < sequence.length; i++) {
    if (sequence[i] !== clean[clean.length - 1]) clean.push(sequence[i]);
  }

  return { sequence: clean, totalDistance: distance };
}

/**
 * Run a named algorithm
 */
export function runAlgorithm(name, requests, initialHead, maxCylinder, direction) {
  switch (name) {
    case 'FCFS':
      return fcfs(requests, initialHead);
    case 'SSTF':
      return sstf(requests, initialHead);
    case 'SCAN':
      return scan(requests, initialHead, maxCylinder, direction);
    case 'C-SCAN':
      return cscan(requests, initialHead, maxCylinder, direction);
    default:
      return fcfs(requests, initialHead);
  }
}
